akka-http/src/main/scala/akka/http/scaladsl/server/directives/FramedEntityStreamingDirectives.scala

```diff
@@ -0,0 +1,239 @@
+   *
+   * See also [[MiscDirectives.withoutSizeLimit]] as you may want to allow streaming infinite streams of data in this route.
+   * By default the uploaded data is limited by the `akka.http.parsing.max-content-length`.
+   */
+  final def asSourceOf[T](implicit um: Unmarshaller[ByteString, T], framing: FramingWithContentType): RequestToSourceUnmarshaller[T] =
+    asSourceOfAsync(1)(um, framing)
+
+  /**
+   * Extracts entity as [[Source]] of elements of type `T`.
+   * This is achieved by applying the implicitly provided (in the following order):
+   *
+   * - 1st: [[FramingWithContentType]] in order to chunk-up the incoming [[ByteString]]s according to the
+   *        `Content-Type` aware framing (for example, [[akka.http.scaladsl.server.EntityStreamingSupport.bracketCountingJsonFraming]]).
+   * - 2nd: [[Unmarshaller]] (from [[ByteString]] to `T`) for each of the respective "chunks" (e.g. for each JSON element contained within an array).
+   *
+   * The request will be rejected with an [[akka.http.scaladsl.server.UnsupportedRequestContentTypeRejection]] if
+   * its [[ContentType]] is not supported by the used `framing` or `unmarshaller`.
+   *
+   * It is recommended to use the [[akka.http.scaladsl.server.EntityStreamingSupport]] trait in conjunction with this
+   * directive as it helps provide the right [[FramingWithContentType]] and [[SourceRenderingMode]] for the most
+   * typical usage scenarios (JSON, CSV, ...).
+   *
+   * Cancelling extracted [[Source]] closes the connection abruptly (same as cancelling the `entity.dataBytes`).
+   *
+   * See also [[MiscDirectives.withoutSizeLimit]] as you may want to allow streaming infinite streams of data in this route.
+   * By default the uploaded data is limited by the `akka.http.parsing.max-content-length`.
+   */
+  final def asSourceOf[T](framing: FramingWithContentType)(implicit um: Unmarshaller[ByteString, T]): RequestToSourceUnmarshaller[T] =
+    asSourceOfAsync(1)(um, framing)
+
+  /**
+   * Similar to [[asSourceOf]] however will apply at most `parallelism` unmarshallers in parallel.
+   *
+   * The source elements emitted preserve the order in which they are sent in the incoming [[HttpRequest]].
+   * If you want to sacrivice ordering in favour of (potential) slight performance improvements in reading the input
+   * you may want to use [[asSourceOfAsyncUnordered]] instead, which lifts the ordering guarantee.
+   *
+   * Refer to [[asSourceOf]] for more in depth-documentation and guidelines.
+   *
+   * See also [[MiscDirectives.withoutSizeLimit]] as you may want to allow streaming infinite streams of data in this route.
+   * By default the uploaded data is limited by the `akka.http.parsing.max-content-length`.
+   */
+  final def asSourceOfAsync[T](parallelism: Int)(implicit um: Unmarshaller[ByteString, T], framing: FramingWithContentType): RequestToSourceUnmarshaller[T] =
```

Comment:
The AsyncUnordered is more interesting in fact, but for completeness keeping both.
It's when some elements are very large and some very small, and you can push out the smaller ones while the big one is still being rendered.

Just async is also OK but not that interesting